### PR TITLE
Start building some checking of diagnostic use.

### PR DESCRIPTION
**Pull request by jonmeow**
In theory we're doing a central registry so that we can ensure there's at least one test for each. This isn't doing that, but I'm trying to validate that the central registry isn't leading to duplicates or abandoned checks (and catching a couple of each).

### CODE REVIEW

File: toolchain/diagnostics/check_diagnostics.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    line_offset = 0
+
+    found: Dict[str, List[Location]] = {}
+    for m in re.finditer(r"CARBON_DIAGNOSTIC\(\s*(\w+),", content):
+        diag = m.group(1)
+        if diag in IGNORED:
+            continue
+        line += content.count("\n", line_offset, m.start())
+        line_offset = m.start()
+        found.setdefault(diag, []).append(Location(path, line))
```
Comment (Pixep) [inline toolchain/diagnostics/check_diagnostics.py]: Would a `defaultdict` make more sense instead of a `Dict`+`setdefault`?

**Review by Pixep — APPROVED** [review]
Great idea, not much to suggest :)

File: toolchain/diagnostics/check_diagnostics.py (hunk 1 above)
Comment (jonmeow) [reply toolchain/diagnostics/check_diagnostics.py]: Sure, done

**Review by chandlerc — APPROVED** [review]
Nice!

I thought about ways we could enforce this in C++, but none seemed really worthwhile... Even if they worked, the error would be much less helpful than this one is. =]

**Pull request merged**